dp/generate_climos.py

```diff
@@ -39,6 +39,86 @@
 cdo = Cdo()
 
 
+def has_climo_periods(input_file, climo):
+    '''
+    This function checks if the given input NetCDF dataset has a variable "climo_periods".
+    generate_climos requires the input dataset to have the variable, thus this function will
+    warn if there is no such variable.
+    '''
+    periods = input_file.climo_periods.keys() & climo
+    logger.info("climo_periods: {}".format(periods))
+    if len(periods) == 0:
+        logger.info(f"{input_file.filepath()} has no variable 'climo_periods'")
+    return periods
+
+
+def is_NetCDF(filepath):
+    '''
+    This function checks if the given input file path is a vaild NetCDF file. It returns
+    NetCDF dataset of the input file. Ohterwise, it halts the program.
+    '''
+    try:
+        return CFDataset(filepath)
+    except Exception as e:
+        logger.info("{}: {}".format(e.__class__.__name__, e))
+        sys.exit()
+
+
+def dry_run_handler(filepath, climo):
+    '''
+    This function handles dry-run operation of generate_climos. The purpose of this function
+    is to check variables and attrbutes to be used for generate_climos. dry-run will not 
+    produca any output files.
+    '''
+    logger.info("")
+    logger.info("File: {}".format(filepath))
+    input_file = is_NetCDF(filepath)
+    has_climo_periods(input_file, climo)
+
```

Comment:
A general rule of coding is DRY - Don't Repeat Yourself. In this method and in `generate_climos`, you have repeated  the same sequence of code: log some info (the same info), open the file, get the climo periods. This should be factored out into the method that calls these two functions, with arguments as appropriate for the results. (You can see why this was rolled up into a single method before.) 

(There's no benefit to slightly changing what is logged between dry run and full run; in fact it is better if they are as close as possible.)

With this in mind, we see that function `is_NetCDF` becomes unnecessary, and its problems (calling `sys.exit()` too far down the hierarchy) can be fixed by putting its work higher up the calling hierarchy.

The same is probably also true of `has_climo_periods`; that is, it is unnecessary given that it would only be called once.

This refactoring is definitely on the right track, but the repetition these two methods indicates a need to reorganize it a little.